### PR TITLE
Feature/account terms

### PR DESCRIPTION
**Pull request by sstruzik**
### Support Account terms
Support terms AccDed6All, AccLimit6All, AccMinDed6All, AccMaxDed6All, AccDedType6All, AccLimitType6All from OED AccAll level
At Account level, policies are applied to the sum of all layers then back allocated to the layer and to the items.

### CODE REVIEW

**Comment by johcarter** [comment]
@sstruzik I've fixed ODS validation issues with the insurance_accounts test so that it can be run on your branch.  I'll let you know of any issues I find with numerical results as they arise

**Review by johcarter — APPROVED** [review]
Numbers look good, just need to add insurance_account fm validation test to ci, ideally with the DRs 0.25 0.5 0.75 1

File: oasislmf/pytools/fm/back_allocation.py [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+        loss_in: loss before applying profile
+        loss_out: loss after applying profile
+        temp_node_loss: array to store loss factor
+        loss_indptr: index to loss pointer
+        loss_val: loss values
+        extra: extra after applying profile
+        temp_node_extras: extra after applying profile(dense)
+        extras_indptr: index to extra pointer
+        extras_val: extra values
+    """
```
Comment (mtazzari) [inline oasislmf/pytools/fm/back_allocation.py:31]: Missing data types of all arguments.

File: oasislmf/pytools/fm/back_allocation.py [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+        node_sidx: sidx for this node
+        sidx_indptr: index to sidx pointer
+        sidx_indexes: index of sidx for nodes
+        sidx_val: sidx values
+        loss_in: loss before applying profile
+        loss_out: loss after applying profile
+        temp_node_loss: array to store loss factor
+        loss_indptr: index to loss pointer
+        loss_val: loss values
+
```
Comment (mtazzari) [inline oasislmf/pytools/fm/back_allocation.py:157]: Missing data types of all arguments.

File: oasislmf/pytools/fm/back_allocation.py [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+            # print('ba', child['level_id'], child['agg_id'], p, loss_indptr[child['loss'] + p], child_loss[0], temp_node_loss[p, -3],
+            # extras_indptr[child['extra'] + p], child_extra[0])
+
+
+@njit(cache=True, fastmath=True)
+def back_alloc_a2(base_children_len, temp_children_queue, nodes_array, p,
+                  node_val_len, node_sidx, sidx_indptr, sidx_indexes, sidx_val,
+                  loss_in, loss_out, temp_node_loss, loss_indptr, loss_val):
+    """
+    back allocation of loss to the base children
```
Comment (mtazzari) [inline oasislmf/pytools/fm/back_allocation.py:141]: The function doesn't return anything, so I presume it modifies some arrays in place.
I think it'd be good to make it clear in the docstring, like 
```suggestion
    back allocation of loss to the base children. The function modifies in-place array xyz.
```

File: oasislmf/pytools/fm/back_allocation.py [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+        node_extra_indptr: index of the extra pointer of the node
+        loss_in: loss before applying profile
+        loss_out: loss after applying profile
+        loss_indptr: index to loss pointer
+        loss_val: loss values
+        temp_node_loss_layer_ba: sparse array to loss after layer back alloc
+        extras_indptr: index to extra pointer
+        extras_val: extra values
+        temp_node_extras_layer_merge: node extra after profile
+        temp_node_extras_layer_merge_save: node extra before profile
```
Comment (mtazzari) [inline oasislmf/pytools/fm/back_allocation.py:234]: missing all data types

File: oasislmf/pytools/fm/back_allocation.py [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+
+@njit(cache=True, fastmath=True)
+def back_alloc_layer_extra(layer_len, node_val_len, node_loss_indptr, node_extra_indptr,
+                           loss_in, loss_out, loss_indptr, loss_val,
+                           temp_node_loss_layer_ba,
+                           extras_indptr, extras_val,
+                           temp_node_extras_layer_merge, temp_node_extras_layer_merge_save
+                           ):
+    """
+
```
[the remark this answers is not in the document]
Comment (mtazzari) [reply oasislmf/pytools/fm/back_allocation.py:220]: The function doesn't return anything, perhaps mention in the docstring which arrays are updated in-place.

File: oasislmf/pytools/fm/compute_sparse.py [hunk 6]
```diff
@@ -1,20 +1,28 @@
...
 
 @njit(cache=True)
 def get_base_children(node, children, nodes_array, temp_children_queue):
+    """
+    fill up temp_children_queue with all the base children of node from index 0 and return the number of base children
+    Args:
+        node: top node
+        children: array of all the children for each node
+        nodes_array: array of information on all nodes
+        temp_children_queue: empty array where we write base children.
```
Comment (mtazzari) [inline oasislmf/pytools/fm/compute_sparse.py:21]: missing data types

File: oasislmf/pytools/fm/compute_sparse.py [hunk 7]
```diff
@@ -1,20 +1,28 @@
...
+
+    Returns:
+        number of base children
```
Comment (mtazzari) [inline oasislmf/pytools/fm/compute_sparse.py:24]: missing return data type

File: oasislmf/pytools/fm/compute_sparse.py [hunk 8]
```diff
@@ -73,29 +80,26 @@ def first_time_layer_extra(profile_len, base_children_len, temp_children_queue,
...
 @njit(cache=True, fastmath=True)
-def aggregate_children_extras(node, len_children, nodes_array, children, temp_children_queue,
-                              temp_node_sidx, sidx_indexes, sidx_indptr, sidx_val, sidx_ptr_i, sidx_i, all_sidx,
-                              temp_node_loss, loss_indptr, loss_val, loss_ptr_i,
-                              temp_node_extras, extras_indptr, extras_val, extras_ptr_i):
+def aggregate_children_extras(node, len_children, nodes_array, children, temp_children_queue, compute_idx,
+                              temp_node_sidx, sidx_indexes, sidx_indptr, sidx_val, all_sidx,
+                              temp_node_loss, loss_indptr, loss_val,
+                              temp_node_extras, extras_indptr, extras_val):
     sidx_created = False
```
Comment (mtazzari) [inline oasislmf/pytools/fm/compute_sparse.py:98]: missing docstring, arguments etc

File: oasislmf/pytools/fm/compute_sparse.py [hunk 9]
```diff
@@ -127,48 +132,49 @@ def aggregate_children_extras(node, len_children, nodes_array, children, temp_ch
...
 
 
 @njit(cache=True, fastmath=True)
-def aggregate_children(node, len_children, nodes_array, children, temp_children_queue,
-                       temp_node_sidx, sidx_indexes, sidx_indptr, sidx_val, sidx_ptr_i, sidx_i, all_sidx,
-                       temp_node_loss, loss_indptr, loss_val, loss_ptr_i):
+def aggregate_children(node, len_children, nodes_array, children, temp_children_queue, compute_idx,
+                       temp_node_sidx, sidx_indexes, sidx_indptr, sidx_val, all_sidx,
+                       temp_node_loss, loss_indptr, loss_val):
     sidx_created = False
```
Comment (mtazzari) [inline oasislmf/pytools/fm/compute_sparse.py:173]: missing docstring, arguments, etc

File: oasislmf/pytools/fm/compute_sparse.py [hunk 10]
```diff
@@ -178,193 +184,55 @@ def aggregate_children(node, len_children, nodes_array, children, temp_children_
...
         loss_indptr[node['loss'] + layer] = loss_indptr[node['loss']]
 
-    return node_val_len, sidx_i, sidx_ptr_i, loss_ptr_i
+    return node_val_len
 
 
 @njit(cache=True)
-def set_parent_next_compute(parent_id, child_id, nodes_array, children, next_compute_i, computes):
+def set_parent_next_compute(parent_id, child_id, nodes_array, children, computes, compute_idx):
     parent = nodes_array[parent_id]
```
Comment (mtazzari) [inline oasislmf/pytools/fm/compute_sparse.py:229]: missing docstring, arguments etc

File: oasislmf/pytools/fm/compute_sparse.py [hunk 11]
```diff
@@ -376,7 +244,7 @@ def compute_event(compute_info,
                   len_array, max_sidx_val, sidx_indexes, sidx_indptr, sidx_val, loss_indptr, loss_val, extras_indptr, extras_val,
                   children,
                   computes,
-                  next_compute_i,
+                  compute_idx,
                   item_parent_i,
                   fm_profile,
                   stepped):
```
Comment (mtazzari) [inline oasislmf/pytools/fm/compute_sparse.py:250]: missing docstring, arguments, etc

File: oasislmf/pytools/fm/compute_sparse.py [hunk 12]
```diff
@@ -389,17 +257,36 @@ def compute_event(compute_info,
     # loss_val = computation_structure.loss_val
     # extras_indptr = computation_structure.extras_indptr
```
Comment (mtazzari) [inline oasislmf/pytools/fm/compute_sparse.py:258]: I'd clean up all this unneeded stuff

**Review by mtazzari — APPROVED** [review]
I see that Joh signed it off, so I understand it the PR produces the desired results.
I have no code comments, just docstrings are missing in many places.
I've added comments in some but not all the functions that lack docstrings.
In some places the entire docstring is missing, in some places just the argument types, etc.

**Review by sambles — CHANGES_REQUESTED** [review]
Only one minor nitpick, in the expected results there are `*.npy` added in this PR

 under [validation/insurance_account/expected/output](https://github.com/OasisLMF/OasisLMF/tree/feature/account_terms/validation/insurance_account/expected/output)

I'd remove those files: (unless needed for something?) 
* [fm_profile.npy](https://github.com/OasisLMF/OasisLMF/blob/feature/account_terms/validation/insurance_account/expected/output/fm_profile.npy)
* [node_parents_array_2.npy](https://github.com/OasisLMF/OasisLMF/blob/feature/account_terms/validation/insurance_account/expected/output/node_parents_array_2.npy)
* [node_profiles_array_2.npy](https://github.com/OasisLMF/OasisLMF/blob/feature/account_terms/validation/insurance_account/expected/output/node_profiles_array_2.npy)
* [nodes_array_2.npy](https://github.com/OasisLMF/OasisLMF/blob/feature/account_terms/validation/insurance_account/expected/output/nodes_array_2.npy)
* [output_array_2.npy](https://github.com/OasisLMF/OasisLMF/blob/feature/account_terms/validation/insurance_account/expected/output/output_array_2.npy)

And add the `*.npy` extension to the `.gitignore` file

**Pull request merged**